### ISSUE TITLE
LaserScan Msgs not received in Unity Although the /scan topic has one publisher: gazebo and one subscriber: rosbridge websocket

### ISSUE THREAD

**Issue by ZahraaBass** [siemens/ros-sharp#372]
* [x] I am at the right place and my issue is directly related to ROS#. General technical questions I would post e.g. at [ROS Answers](https://answers.ros.org/) or [Stack Overflow](https://stackoverflow.com). For library-specific questions I would look for help in the corresponding library forums.
* [x] I have thoroughly read [the Contributing Guideline](Contributing.md) and writing this issue is the right thing to do in my case.

---
## I have a question! ##

* [x] I searched the [Wiki](https://github.com/siemens/ros-sharp/wiki), [open](https://github.com/siemens/ros-sharp/issues) and [closed](https://github.com/siemens/ros-sharp/issues?q=is%3Aissue+is%3Aclosed) issues for an answer. I tried my best to find the answer by myself without success. I believe that the discussion we will have in this issue, and the solutions we might find, will help me, and likely other community members who have a similar problem.

**Here is my question:**

Hello, I'm facing a problem with receiving LaserScan msgs. In my case I was able to receive fake odometry data from ROS through a talker node publishing on the /chatter_ topic. A link showing what I did: https://www.youtube.com/watch?v=yXWNSwk4Irs
I was also able to import the turtlebot URDF File: https://www.youtube.com/watch?v=jSc0PF_tSAQ

However, when I use the LaserScanSubscriber, and subscribe to the /scan topic, the DebugLog print in ReceiveMessage are not printing on the Console. Basically what I'm doing is launching the husky gazebo simulation, which publishes Lidar Data on the /scan topic like ranges. On the ROS side, the /scan topic shows one publisher (gazebo) and one subscriber (Rosbridge). so the problem is on the Unity side.. I've been trying to solve this problem for more than a week and I'm desperate for some clue. I'm kinda a newbie to both (ROS and Unity) and would really appreciate some help/insight.

I made a video documenting all the steps I took and what my problem is: https://www.youtube.com/watch?v=L9aU0XW6G1o

I tried several things to solve this, for example to click on RosbridgeClient - Auto Generate Messages - Simple Message -Brows File , but here when I go to the msg folder inside the MessageTypes Folder and try to click on anything, nothing can be clicked.

<img width="1102" alt="Screen Shot 2021-01-21 at 9 22 49 AM" src="https://user-images.githubusercontent.com/76676384/105318512-4be9c680-5bcc-11eb-96bf-67876dbed23b.png">

Another thing I tried to drag and drop the LaserScan.cs file into the Assets folder in Unity Project, but then these errors will come out: 
Assets/RosSharp/Scripts/RosBridgeClient/RosCommuncation/LaserScanPublisher.cs(48,47): warning CS0436: The type 'LaserScan' in '/Users/zahraabassyouni/Unity/Projects/ROS&Unity/Assets/LaserScan.cs' conflicts with the imported type 'LaserScan' in 'RosBridgeClient, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. Using the type defined in '/Users/zahraabassyouni/Unity/Projects/ROS&Unity/Assets/LaserScan.cs'.

<img width="1430" alt="Screen Shot 2021-01-21 at 9 38 14 AM" src="https://user-images.githubusercontent.com/76676384/105318604-64f27780-5bcc-11eb-8c17-7a310bb611ae.png">

**Comment by ZahraaBass** [comment]
update: it turned out that I rosbridge_suite installed...
so now I can chose a type of msg and ROS# will autogenerate 
<img width="1535" alt="Screen Shot 2021-01-22 at 8 02 27 AM" src="https://user-images.githubusercontent.com/76676384/105453091-2cb06f00-5c88-11eb-9220-a769cabfa4f2.png">

but , still, the Debug.Log line in the ReceiveMessage function of LaserScan Subscriber aren't printing

**Comment by ZahraaBass** [comment]
okay i guess I located what the exact problem is.
<img width="870" alt="Screen Shot 2021-01-22 at 3 46 57 PM" src="https://user-images.githubusercontent.com/76676384/105498608-11198880-5cc9-11eb-849c-11ea049b9ffa.png">

The Header in LaserScan.cs is not colored, which means that VS cannot know where the class Header is located...
Only when I open Header.cs and play, the Header word gets colored...
<img width="868" alt="Screen Shot 2021-01-22 at 3 49 02 PM" src="https://user-images.githubusercontent.com/76676384/105498875-5b9b0500-5cc9-11eb-894f-1c662d4e2238.png">

this means that VS isn't reading msgs which are ros specific, but still i dunno how to solve it.

**Comment by berkayalpcakal** [comment]
Hi @ZahraaBass ,

Sorry that I am a bit confused. When you put all the ROS# scripts together with message classes into Unity, do you get a build error, or is the situation as it is in the video link above still valid?

As an alternative way of debugging, you can always import the source code of the [RosBridgeClient](https://github.com/siemens/ros-sharp/tree/master/Libraries/RosBridgeClient) into your Assets instead of importing the [.dll](https://github.com/siemens/ros-sharp/blob/master/Unity3D/Assets/RosSharp/Plugins/RosBridgeClient.dll) file, so that you can fully debug the code.  I suggest you to check the line regarding the [deserialized data](https://github.com/siemens/ros-sharp/blob/master/Libraries/RosBridgeClient/RosSocket.cs#L180) received from rosbridge, which includes error logs if any.

**Comment by ZahraaBass** [comment]
Hello @berkayalpcakal, Thank you so much for replying.

I'm sorry for causing this confusion. This [link](url) (https://www.youtube.com/watch?v=L9aU0XW6G1o) is still valid, so no I wasn't getting any error during build. Another update is that I was able to get the ranges from ROS but like one value at a time. So when the msg is just one number Unity seems to have no problem in the ReceiverMessage function, but when the msg is made up of more than one entity, the ReceiveMessage function isnt called - like doesnt even execute a Debug.Log("Hello");

Regarding the alternative way you mentioned, you mean putting Libraries directly into the Assets folder instead of the plugins, like what I did here: https://www.youtube.com/watch?v=pHOIiGMqkMM&feature=youtu.be
I'm not sure if I did it right because I got alot of errors (shown in the video).

And one thing I noticed is that the Time.cs class is defined in the namespace RosSharp.RosBridgeClient.MessageTypes.Std although its not in the MessageTypes folder but in the BaseMessages folder. 

<img width="1050" alt="Screen Shot 2021-01-24 at 11 49 59 PM" src="https://user-images.githubusercontent.com/76676384/105644742-d4ec5080-5e9f-11eb-9d76-516f194c9a79.png">

**Comment by berkayalpcakal** [comment]
Hi again @ZahraaBass ,

What I understood from the video is that you placed the library source scripts together with the library dll's, which leads to the errors regarding the duplicate definition of the classes. Instead of  copying and pasting the entire library folder, please try placing only the .cs source scripts into the assets folder.

**Comment by ZahraaBass** [comment]
Hello @berkayalpcakal, thanks for the update.
okay I think I did it almost right this time: https://youtu.be/27q1Np9LjOo
I still have 24 errors which are mainly caused by the AssemblyInfo.cs file (there are three files with the same name), for example the errors are like: 
_Assets/Libraries/RosBridgeClient/Properties/AssemblyInfo.cs(31,12): error CS0579: Duplicate 'AssemblyCulture' attribute_

If I delete the AssemblyInfo.cs files in Assets/Libraries/Urdf/Properties and Assets/Libraries/MessageGeneration/Properties, that solved the errors but not sure if this is the right way? Also this would generate  this warning (but it can be cleared):
_Assets/Libraries/RosBridgeClient/Protocols/WebSocketNetProtocol.cs(108,26): warning CS1702: Assuming assembly reference 'System.Threading.Tasks.Extensions, Version=4.2.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' used by 'System.Threading.Channels' matches identity 'System.Threading.Tasks.Extensions, Version=4.2.0.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' of 'System.Threading.Tasks.Extensions', you may need to supply runtime policy_

**Comment by ZahraaBass** [comment]
I have an update: https://youtu.be/ILsqY7fEucU

I solved the errors (as I mentioned above) and was able to show the msg on the unity console through adding a Debug.Log line in Rossocket.cs script (as shown in video).

**Comment by berkayalpcakal** [comment]
Hello again @ZahraaBass,

I am glad to hear that you made a progress. Could you please write down what went wrong before, and how you figured it out?
If you think that the issue is resolved, please close the issue.

**Comment by ZahraaBass** [comment]
first changed the API compatability levelto .Net 4.x.
 (the upper tab, press on Edit – Project Settings – player – other settings, and change the API compatability level to : .Net 4.x)
Then  imported the source code of the RosBridgeClient into my Assets instead of importing the .dll file as in this [video](https://youtu.be/27q1Np9LjOo)
this resulted in some errors (which I mentioned how I solved in the previous comment).
and this [video](https://youtu.be/ILsqY7fEucU) shows where I added the Debug.Log line in Rossocket.cs script.
and Thank you so much for the help! @berkayalpcakal

**Comment by ZahraaBass** [comment]
Thanks to issue #377, I found what the real problem was. The LaserScan messages coming from Husky in the PlaypenWorld had some ranges which were 'inf'. So I wrote a node that subscribes to /scan topic to get the LaserScan msgs, changes the 'inf' values to 30 (max range of husky robot) and publishes on a topic (/chatter_). Screenshot below.

<img width="754" alt="Screen Shot 2021-02-03 at 3 10 47 PM" src="https://user-images.githubusercontent.com/76676384/106751549-00a2cf80-6632-11eb-8a3d-29713495a843.png">

Now the LaserScan Subscriber works and the LaserWriter shows these values in the Inspector window
Ofcourse the solution is mentioned in issue #78 but it seems that I didn't understand the title (new to this JSON serialization thing). Anyways thanks!
